eZ/Publish/SPI/FieldType/Generic/Type.php

```diff
@@ -0,0 +1,360 @@
+     * value in either sort_key_string or sort_key_int.
+     *
+     * In case of multi value, values should be string and separated by "-" or ",".
+     *
+     * @param \eZ\Publish\Core\FieldType\Value $value
+     *
+     * @return mixed
+     */
+    protected function getSortInfo(Value $value)
+    {
+        return null;
+    }
+
+    public function toPersistenceValue(Value $value): PersistenceValue
+    {
+        return new PersistenceValue(
+            [
+                'data' => $this->toHash($value),
+                'externalData' => null,
+                'sortKey' => $this->getSortInfo($value),
+            ]
+        );
+    }
+
+    public function fromPersistenceValue(PersistenceValue $fieldValue)
+    {
+        return $this->fromHash($fieldValue->data);
+    }
+
+    public function isSearchable(): bool
+    {
+        return false;
+    }
+
+    public function isSingular(): bool
+    {
+        return false;
+    }
+
+    public function onlyEmptyInstance(): bool
+    {
+        return false;
+    }
+
+    public function isEmptyValue(Value $value): bool
+    {
+        return $value === null || $value == $this->getEmptyValue();
+    }
+
+    final public function acceptValue($inputValue): Value
+    {
+        if ($inputValue === null) {
+            return $this->getEmptyValue();
+        }
+
+        $value = $this->createValueFromInput($inputValue);
+
+        $this->checkValueType($value);
+
+        if ($this->isEmptyValue($value)) {
+            return $this->getEmptyValue();
+        }
+
+        return $value;
+    }
+
+    /**
+     * Inspects given $inputValue and potentially converts it into a dedicated value object.
+     *
+     * If given $inputValue could not be converted or is already an instance of dedicate value object,
+     * the method should simply return it.
+     *
+     * This is an operation method for {@see acceptValue()}.
+     *
+     * Example implementation:
+     * <code>
+     *  protected function createValueFromInput( $inputValue )
+     *  {
+     *      if ( is_array( $inputValue ) )
+     *      {
+     *          $inputValue = \eZ\Publish\Core\FieldType\CookieJar\Value( $inputValue );
+     *      }
+     *
+     *      return $inputValue;
+     *  }
+     * </code>
+     *
+     * @param mixed $inputValue
+     *
+     * @return mixed The potentially converted input value.
+     */
+    protected function createValueFromInput($inputValue)
+    {
+        if (is_string($inputValue)) {
+            $inputValue = $this->serializer->denormalize(
+                $this->serializer->decode($inputValue),
+                $this->getValueClass()
+            );
+        }
+
+        return $inputValue;
+    }
+
+    /**
+     * Returns FQN of class representing Field Type Value.
+     *
+     * @return string
+     */
+    protected function getValueClass(): string
+    {
+        return substr_replace(static::class, 'Value', strrpos(static::class, '\\') + 1);
```

Comment:
FWIW this kind of check for a Value instance was a mistake from a while ago; in case you need to override the Type it also forces you to override the value, which can be a real pain if somewhere in your codebase the original Value is used. If would have been better if each field type had its own implementation instead.

Reply:
Hi @pspanja! Thank you for your feedback!

> in case you need to override the Type it also forces you to override the value, which can be a real pain if somewhere in your codebase the original Value is used.

This is a common issue with the "convention over configuration" approach. 

> It would have been better if each field type had its own implementation instead.

The use case which you mentioned seems to be very rare. Is it good enough reason for forcing everyone to implement "getValueClass"?

Reply:
It should, at very least be public, so it can be overriden in a decorator to restore the old value if the type is overriden.

Reply:
> This method should, at very least, be public and part of the interface, so it can be overriden in a decorator to restore the old value if the type is overriden.

+1

Reply:
Sounds reasonable. Thank you for suggestion @emodric 😉

Reply:
Note: @adamwojs will provide follow-up with this shortly. It will be easier to review. I'm merging this PR :)